### PR TITLE
Extract into crates

### PR DESCRIPTION
**Pull request by wycats**
The key commit in this PR is https://github.com/nushell/nushell/commit/e4226def162012e7fafc3f74242cdf290ce19113

---

This commit extracts five new crates:

- nu-source, which contains the core source-code handling logic in Nu, including Text, Span, and also the pretty.rs-based debug logic
- nu-parser, which is the parser and expander logic
- nu-protocol, which is the bulk of the types and basic conveniences used by plugins
- nu-errors, which contains ShellError, ParseError and error handling conveniences
- nu-textview, which is the textview plugin extracted into a crate

One of the major consequences of this refactor is that it's no longer possible to `impl X for Spanned<Y>` outside of the `nu-source` crate, so a lot of types became more concrete (Value became a concrete type instead of Spanned<Value>, for example).

This also turned a number of inherent methods in the main nu crate into plain functions (impl Value {} became a bunch of functions in the `value` namespace in `crate::data::value`).

---

The rest of the PR is mostly about keeping the branch up to date with master.

### CODE REVIEW

File: docs/commands/config.md [hunk 1]
```diff
@@ -30,14 +29,25 @@ Syntax: `config {flags}`
 
 ### Variables
 
-| Variable        | Type           | Description  |
+<<<<<<< HEAD
```
Comment (jonathandturner) [inline docs/commands/config.md]: This seems bad

File: src/commands/classified/external.rs [hunk 2]
```diff
@@ -43,233 +47,163 @@ impl Decoder for LinesCodec {
...
-                });
-
-                format!("{} {}", self.name, itertools::join(args, " "))
-            });
+                })
+            })
+            .collect::<Result<Vec<String>, ShellError>>()?;
 
-            process = Exec::shell(itertools::join(commands, " && "))
-        } else {
```
Comment (jonathandturner) [inline src/commands/classified/external.rs]: This feels a little surprising, but I'd need to try it out

File: src/commands/classified/pipeline.rs [hunk 3]
```diff
@@ -1,107 +1,74 @@
...
+            (Some(ClassifiedCommand::Internal(left)), _) => {
+                let stream = run_internal_command(left, ctx, input, Text::from(line)).await?;
+                ClassifiedInputStream::from_input_stream(stream)
+            }
 
-                (Some(ClassifiedCommand::External(left)), Some(ClassifiedCommand::External(_))) => {
-                    left.run(ctx, input, StreamNext::External).await?
-                }
+            (Some(ClassifiedCommand::External(left)), Some(ClassifiedCommand::External(_))) => {
+                run_external_command(left, ctx, input, StreamNext::External).await?
```
Comment (jonathandturner) [inline src/commands/classified/pipeline.rs]: I'm a little confused about this bit. I think the intent of the PR this came from was to begin to consolidate the command types more. This seems to segregate with different helper commands.

File: src/commands/classified/external.rs (hunk 2 above)
Comment (wycats) [reply src/commands/classified/external.rs]: I'm not sure it was intentional. Are there tests for the original change?

File: src/commands/classified/pipeline.rs (hunk 3 above)
Comment (wycats) [reply src/commands/classified/pipeline.rs]: the `.run` methods were previously inherent methods on the two command types, but the command types were moved into the parser. In general, when there were only a handful of uses of an inherent method, I addressed these issues for now by moving the inherent methods to functions.

File: src/commands/classified/external.rs (hunk 2 above)
Comment (jonathandturner) [reply src/commands/classified/external.rs]: Not that I know of. Can we just use what was there before? We can focus on changing it in a more focused PR later.

**Pull request merged**